pkg/utils/gwlog/metadata.go

```diff
@@ -0,0 +1,85 @@
+package gwlog
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+)
+
+const metadataKey string = "metadata_key"
+const traceID string = "trace_id"
+
+type metadata struct {
+	m map[string]string
+}
+
+func (mv *metadata) get(key string) string {
+	return mv.m[key]
+}
+
+func (mv *metadata) set(key, val string) {
+	mv.m[key] = val
+}
+
+func newMetadata() *metadata {
+	return &metadata{
+		m: make(map[string]string),
+	}
+}
+
+func NewTrace(ctx context.Context) context.Context {
+	currID := uuid.New()
+
+	newCtx := context.WithValue(ctx, metadataKey, newMetadata())
+	AddMetadata(newCtx, traceID, currID.String())
+
+	return newCtx
+}
+
+func AddMetadata(ctx context.Context, key, value string) {
+	if ctx.Value(metadataKey) != nil {
+		ctx.Value(metadataKey).(*metadata).set(key, value)
+	}
+}
+
+func getMetadata(ctx context.Context) []interface{} {
```

Comment:
Just leaving a thought for later consideration. I think rather than computing this dynamically, it might make more sense from a performance perspective to simply keep a copy of this ready whenever a mutation is made to the underlying map.